File: 1_카카오/프로그래머스Lv3-외벽 점검.js
function permute(array, k) {
  if (k > array.length) return null
  const results = []
  prevElements = []
  function dfs(elements, k) {
    if (k === 0) {
      results.push([...prevElements])
      return
    }

    for (let i = 0; i < elements.length; i++) {
      nextElements = [...elements]
      nextElements.splice(i, 1)

      prevElements.push(elements[i])
      dfs(nextElements, k - 1)
      prevElements.pop()
    }
  }
  dfs(array, k)
  return results
}

function solution(n, weak, dist) {
  const W_LENGTH = weak.length
  const PEOPLE_LENGTH = dist.length
  const weaks = [...weak]
  weak.forEach((w) => weaks.push(w + n))

  const permutations = permute(dist, PEOPLE_LENGTH)

  let result = 999

  for (let i = 0; i < W_LENGTH; i++) {
    for (const people of permutations) {
      let count = 1
      let distance = weaks[i] + people[count - 1]
      for (let j = i; j < i + W_LENGTH; j++) {
        if (distance < weaks[j]) {
          count++
          if (count > PEOPLE_LENGTH) break
          distance = weaks[j] + people[count - 1]
        }
      }
      result = Math.min(result, count)
    }
  }

  return result > PEOPLE_LENGTH ? -1 : result
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.46ms, 33.4MB)
// 테스트 2 〉	통과 (0.44ms, 33.4MB)
// 테스트 3 〉	통과 (55.52ms, 49.3MB)
// 테스트 4 〉	통과 (48.44ms, 49.2MB)
// 테스트 5 〉	통과 (0.97ms, 33.5MB)
// 테스트 6 〉	통과 (16.18ms, 38.7MB)
// 테스트 7 〉	통과 (0.26ms, 33.5MB)
// 테스트 8 〉	통과 (0.70ms, 33.6MB)
// 테스트 9 〉	통과 (0.41ms, 33.4MB)
// 테스트 10 〉	통과 (53.86ms, 49.1MB)
// 테스트 11 〉	통과 (54.64ms, 49.2MB)
// 테스트 12 〉	통과 (55.32ms, 49.1MB)
// 테스트 13 〉	통과 (55.07ms, 49.2MB)
// 테스트 14 〉	통과 (60.96ms, 49.2MB)
// 테스트 15 〉	통과 (56.34ms, 49.2MB)
// 테스트 16 〉	통과 (51.57ms, 49.4MB)
// 테스트 17 〉	통과 (69.47ms, 49.1MB)
// 테스트 18 〉	통과 (70.08ms, 49MB)
// 테스트 19 〉	통과 (75.56ms, 49.2MB)
// 테스트 20 〉	통과 (51.45ms, 48.9MB)
// 테스트 21 〉	통과 (68.20ms, 49.3MB)
// 테스트 22 〉	통과 (0.68ms, 33.4MB)
// 테스트 23 〉	통과 (0.41ms, 33.4MB)
// 테스트 24 〉	통과 (0.48ms, 33.4MB)
// 테스트 25 〉	통과 (39.70ms, 49.2MB)

File: 1_카카오/프로그래머스lv3-외벽 점검.py
from itertools import permutations
def solution(n, weak, dist):
    W_LENGTH = len(weak)
    PEOPLE_LENGTH = len(dist)
    # 취약지점을 2배 늘려준다.
    weaks = weak[:]
    for w in weak:
        weaks.append(w + n)
        
    # 사람들의 순서를 순열로 만든다.
    permus = list(permutations(dist, PEOPLE_LENGTH))
    
    result = 999
    
    for i in range(W_LENGTH):
        for people in permus:
            count = 1
            # 맨 처음 첫번째 사람이 갈 수 있는 거리
            distance = weaks[i] + people[count - 1]
            # 사람을 보내면서 그 다음 사람을 투입
            for j in range(i, i + W_LENGTH):
                if distance < weaks[j]:
                    count += 1
                    if count > PEOPLE_LENGTH:
                        break
                    distance = weaks[j] + people[count - 1]
            result = min(result, count)
    
    return -1 if result > PEOPLE_LENGTH else result

"""
정확성  테스트
테스트 1 〉	통과 (0.06ms, 10.2MB)
테스트 2 〉	통과 (0.05ms, 10.1MB)
테스트 3 〉	통과 (752.53ms, 14.8MB)
테스트 4 〉	통과 (608.83ms, 14.7MB)
테스트 5 〉	통과 (1.62ms, 10.2MB)
테스트 6 〉	통과 (110.42ms, 10.3MB)
테스트 7 〉	통과 (0.05ms, 10.2MB)
테스트 8 〉	통과 (0.49ms, 10.2MB)
테스트 9 〉	통과 (0.23ms, 10.3MB)
테스트 10 〉	통과 (1429.66ms, 14.7MB)
테스트 11 〉	통과 (1100.83ms, 14.8MB)
테스트 12 〉	통과 (1138.58ms, 14.7MB)
테스트 13 〉	통과 (973.18ms, 14.7MB)
테스트 14 〉	통과 (1369.85ms, 14.7MB)
테스트 15 〉	통과 (1328.90ms, 14.8MB)
테스트 16 〉	통과 (1126.87ms, 14.8MB)
테스트 17 〉	통과 (1355.81ms, 14.9MB)
테스트 18 〉	통과 (1178.38ms, 14.7MB)
테스트 19 〉	통과 (1160.65ms, 14.8MB)
테스트 20 〉	통과 (1116.99ms, 14.7MB)
테스트 21 〉	통과 (1082.40ms, 14.8MB)
테스트 22 〉	통과 (0.26ms, 10.1MB)
테스트 23 〉	통과 (0.42ms, 10.2MB)
테스트 24 〉	통과 (0.51ms, 10.2MB)
테스트 25 〉	통과 (323.11ms, 14.7MB)
"""